Refactor build-manifest to minimize the number of changes needed to add a new component

jyn514 opened the pull request:
- Add all components to `PkgType`
- Automate functionality wherever possible, so functions often don't have to be manually edited
- Where that's not possible, use exhaustive matches on `PkgType` instead of adding individual strings.
- Add documentation for how to add a component. Improve the existing documentation for how to test changes.

I tested locally that this generates an identical manifest before and after my change, as follows:
```sh
git checkout d44e14225ab00e164aa9ea9e8d9e1bee40f96b3e
cargo +nightly run --manifest-path src/tools/build-manifest/Cargo.toml build/dist build/manifest-before 1970-01-01 http://example.com nightly
git checkout refactor-build-manifest
cargo +nightly run --manifest-path src/tools/build-manifest/Cargo.toml build/dist build/manifest-before 1970-01-01 http://example.com nightly
sort -u build/manifest-before/channel-rust-nightly.toml | diff - <(sort -u build/manifest-after/channel-rust-nightly.toml)
```
I then verified by hand that the differences before sorting are inconsequential (mostly targets being slightly reordered).

The only change in behavior is that `llvm-tools` is now properly renamed to `llvm-tools-preview`:
```
; sort -u build/manifest-before/channel-rust-nightly.toml | diff - <(sort -u build/manifest-after/channel-rust-nightly.toml)
784a785
> [renames.llvm-tools]
894a896
> to = "llvm-tools-preview"
```

This is based on https://github.com/rust-lang/rust/pull/102241 and should not be merged before.

jyn514 committed:
More build-manifest docs

jyn514 committed:
More build-manifest docs

jyn514 committed:
build-manifest: Add a macro that makes it impossible to typo `-preview`, or have a mismatch between parsing and stringifying

jyn514 committed:
Use `PkgType` to determine which packages to add to the manifest

Previously, these had to be hard-coded (i.e. specified in both `PkgType` and `fn package`). Now they only have to be specified in `PkgType`.

jyn514 committed:
Use `PkgType::is_preview` to determine whether to add a rename to the manifest

This caught a missing preview rename for `llvm-tools`.

jyn514 committed:
Use an exhaustive match in `target_host_combination`.

This avoids bugs where components are added to one part of the manifest but not another.

rust-highfive commented:
r? @Mark-Simulacrum

(rust-highfive has picked a reviewer for you, use r? to override)

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/3167495800/jobs/5158028461) [(plain)](https://github.com/rust-lang/rust/commit/a8138e0a559ee33c6546e79fa4d079b03ea3b3db/checks/8660880945/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
configure: rust.debug-assertions := True
configure: rust.overflow-checks := True
configure: llvm.assertions      := True
configure: dist.missing-tools   := True
configure: build.configure-args := ['--enable-sccache', '--disable-manage-submodu ...
configure: writing `config.toml` in current directory
configure: 
configure: run `python /checkout/x.py --help`
Attempting with retry: make prepare
---
* highest error code: E0790
Found 506 error codes
Found 0 error(s) in error codes
Done!
tidy error: /checkout/src/tools/build-manifest/src/main.rs:449: TODO is deprecated; use FIXME
some tidy checks failed
Build completed unsuccessfully in 0:00:10
```

</details>

Mark-Simulacrum commented:
Do we have a sense of the likely impact of the llvm-tools rename? rustup component add llvm-tools already doesn't work, so I assume this is harmless?

In any case let's get a @bors try build so we can run this past dev-static nightly and verify it all works in production.

bors commented:
:hourglass: Trying commit 59c3e8621cf1a461ca0e84f849443cc6eb95b7f9 with merge aac0f44457bc655e570a51efc55e244d3c772d85...
<!-- homu: {"type":"TryBuildStarted","head_sha":"59c3e8621cf1a461ca0e84f849443cc6eb95b7f9","merge_sha":"aac0f44457bc655e570a51efc55e244d3c772d85"} -->

Mark-Simulacrum commented on src/tools/build-manifest/README.md:
```diff
@@ -1,27 +1,34 @@
+There are several steps involved here.
+1. Add a new `Step` to `dist.rs`. This should usually be named after the filename of the uploaded tarball. See https://github.com/rust-lang/rust/pull/101799/files#diff-2c56335faa24486df09ba392d8900c57e2fac4633e1f7038469bcf9ed3feb871 for an example.
+    a. If appropriate, call `tarball.is_preview(true)` for the component.
+2. Add the component to the `Extended` tarball in `dist.rs`. This step cannot be skipped; without it, the artifacts will be uploaded and the component metadata will be present in the manifest, but rustup will not realize it's available.
```
This feels pretty iffy. Do we really care about the extended tarball in rustup somehow? I don't think we put all components in there today, and I don't know that we want to even if we do.

Can you help me identify where rustup is consuming the extended tarball?

jyn514 replied:
I think I misidentified the problem here; I was missing the new component in the "extensions" section of the `rust` package, which is unrelated to the `Extended` tarball. https://github.com/rust-lang/rust/pull/102241/files#diff-1cb8c241c2567e1ac5beb8021e009f2ab69fdbaed8df0f4787cb87e268399112R446

Mark-Simulacrum commented on src/tools/build-manifest/src/main.rs:
```diff
@@ -413,42 +401,55 @@ impl Builder {
+                    extensions.push(Component::from_str(&pkg.manifest_component_name(), "*"));
+                }
+                PkgType::Rust | PkgType::Other(_) => {}
+                // FIXME: is this correct? maybe we should add it so rustup knows about it ...
```
This is correct, the tarball here is distributed for those trying to verify reproducible builds, but we don't actually expect anyone to use it via rustup (only manual downloads).

Mark-Simulacrum commented on src/tools/build-manifest/src/main.rs:
```diff
@@ -498,6 +499,12 @@ impl Builder {
         targets: &[&str],
         fallback: &[(&str, &str)],
     ) {
+        if pkgname == "rust" {
```
Hm, should we be passing PkgType around rather than a pkgname (string)?

Mark-Simulacrum reviewed (commented):
Broadly speaking this looks like an excellent improvement, thanks! Left some relatively minor comments, hopefully.

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/3211766507/jobs/5250621570)
Build commit: aac0f44457bc655e570a51efc55e244d3c772d85 (`aac0f44457bc655e570a51efc55e244d3c772d85`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/3211766507/jobs/5250621570"},"merge_sha":"aac0f44457bc655e570a51efc55e244d3c772d85"} -->

Mark-Simulacrum commented:
Kicked off a dev-static nightly.

jyn514 committed:
Use PkgType in more places

In particular, this avoids serializing and parsing the pkg to a string,
which allows getting rid of `PkgType::Other` altogether

jyn514 committed:
Use PkgType for profiles, too

This makes it easier to remove `is_preview` from components in the future if we choose to do so.

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/3339524877/jobs/5528579673) [(plain)](https://github.com/rust-lang/rust/commit/1b12dd4eaa76967423a2ed94ea51c9a1d655064c/checks/9147422486/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
configure: rust.debug-assertions := True
configure: rust.overflow-checks := True
configure: llvm.assertions      := True
configure: dist.missing-tools   := True
configure: build.configure-args := ['--enable-sccache', '--disable-manage-submodu ...
configure: writing `config.toml` in current directory
configure: 
configure: run `python /checkout/x.py --help`
Attempting with retry: make prepare
---
skip untracked path cpu-usage.csv during rustfmt invocations
skip untracked path src/doc/book/ during rustfmt invocations
skip untracked path src/doc/rust-by-example/ during rustfmt invocations
skip untracked path src/llvm-project/ during rustfmt invocations
Diff in /checkout/src/tools/build-manifest/src/main.rs at line 285:
     fn add_packages_to(&mut self, manifest: &mut Manifest) {
         for pkg in PkgType::all() {
             let fallback = if pkg.use_docs_fallback() { DOCS_FALLBACK } else { &[] };
-            self.package(
-                pkg,
-                &mut manifest.pkg,
-            );
-            );
+            self.package(pkg, &mut manifest.pkg, fallback);
     }
 
Diff in /checkout/src/tools/build-manifest/src/main.rs at line 325:
 
 
         let mut default = minimal;
         default.extend([HtmlDocs, Rustfmt, Clippy]);
-            "default",
-            &default
-        );
-        );
+        profile("default", &default);
 
         // NOTE: this profile is effectively deprecated; do not add new components to it.
         let mut complete = default;
Diff in /checkout/src/tools/build-manifest/src/main.rs at line 335:
         complete.extend([Rls, RustAnalyzer, RustSrc, LlvmTools, RustAnalysis, Miri]);
-            "complete",
-            &complete
-        );
-        );
+        profile("complete", &complete);
 
         // The compiler libraries are not stable for end users, and they're also huge, so we only
         // `rustc-dev` for nightly users, and only in the "complete" profile. It's still possible
Diff in /checkout/src/tools/build-manifest/src/main.rs at line 401:
                 PkgType::RustStd => {
                     components.push(host_component(pkg));
                     extensions.extend(
-                        TARGETS.iter().filter(|&&target| target != host).map(|target| {
-                            Component::from_pkg(pkg, target)
+                        TARGETS
+                            .iter()
+                            .iter()
+                            .filter(|&&target| target != host)
+                            .map(|target| Component::from_pkg(pkg, target)),
                 }
                 }
                 // so is rust-mingw if it's available for the target
Diff in /checkout/src/tools/build-manifest/src/main.rs at line 425:
                     extensions.push(host_component(pkg));
                 }
                 PkgType::RustcDev | PkgType::RustcDocs => {
-                    extensions.extend(
-                        HOSTS.iter().map(|target| {
-                            Component::from_pkg(pkg, target)
-                    );
-                    );
+                    extensions.extend(HOSTS.iter().map(|target| Component::from_pkg(pkg, target)));
                 }
                 PkgType::RustSrc => {
                     extensions.push(Component::from_pkg(pkg, "*"));
Diff in /checkout/src/tools/build-manifest/src/main.rs at line 467:
         dst: &mut BTreeMap<String, Vec<String>>,
         pkgs: &[PkgType],
     ) {
-        dst.insert(profile_name.to_owned(), pkgs.iter().map(|s| s.manifest_component_name()).collect());
+            profile_name.to_owned(),
+            profile_name.to_owned(),
+            pkgs.iter().map(|s| s.manifest_component_name()).collect(),
     }
 
     fn extend_profile(
Diff in /checkout/src/tools/build-manifest/src/main.rs at line 493:
Diff in /checkout/src/tools/build-manifest/src/main.rs at line 493:
             return;
         }
 
-        let version_info = self
-            .versions
-            .version(&pkg)
-            .expect("failed to load package version");
+        let version_info = self.versions.version(&pkg).expect("failed to load package version");
         let mut is_present = version_info.present;
 
         // Never ship nightly-only components for other trains.
Diff in /checkout/src/tools/build-manifest/src/main.rs at line 535:
             Target::unavailable()
 
 
-        let targets = pkg.targets()
+        let targets = pkg
+            .targets()
             .iter()
             .map(|name| {
                 let target = if is_present {
Diff in /checkout/src/tools/build-manifest/src/manifest.rs at line 1:
 use crate::versions::PkgType;
+use crate::Builder;
+use crate::Builder;
 use serde::{Serialize, Serializer};
 use std::collections::BTreeMap;
 use std::path::{Path, PathBuf};
Running `"/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/rustfmt" "--config-path" "/checkout" "--edition" "2021" "--unstable-features" "--skip-children" "--check" "/checkout/src/tools/lint-docs/src/lib.rs" "/checkout/src/tools/replace-version-placeholder/src/main.rs" "/checkout/src/tools/tier-check/src/main.rs" "/checkout/src/tools/build-manifest/src/main.rs" "/checkout/src/tools/build-manifest/src/checksum.rs" "/checkout/src/tools/build-manifest/src/versions.rs" "/checkout/src/tools/build-manifest/src/manifest.rs" "/checkout/src/tools/lint-docs/src/main.rs"` failed.
If you're running `tidy`, try again with `--bless`. Or, if you just want to format code, run `./x.py fmt` instead.
```

</details>

Mark-Simulacrum commented on src/tools/build-manifest/src/main.rs:
```diff
@@ -287,28 +283,10 @@ impl Builder {
-                self.package($name, &mut manifest.pkg, $targets, &[])
-            };
+        for pkg in PkgType::all() {
+            let fallback = if pkg.use_docs_fallback() { DOCS_FALLBACK } else { &[] };
```
It seems like this might as well go into `self.package(...)`, now that it's just a call to a method on pkg?

jyn514 replied:
Good idea, done.

Mark-Simulacrum commented:
@bors try

Will kick off another dev-static nightly with this.

bors commented:
:hourglass: Trying commit fde05ea0898bf0436f9a85d6f58064cba9c8336c with merge 16d56136c42dabd7da968fcd4a16adc8f159bc6e...
<!-- homu: {"type":"TryBuildStarted","head_sha":"fde05ea0898bf0436f9a85d6f58064cba9c8336c","merge_sha":"16d56136c42dabd7da968fcd4a16adc8f159bc6e"} -->

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/3352573283/jobs/5555414769)
Build commit: 16d56136c42dabd7da968fcd4a16adc8f159bc6e (`16d56136c42dabd7da968fcd4a16adc8f159bc6e`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/3352573283/jobs/5555414769"},"merge_sha":"16d56136c42dabd7da968fcd4a16adc8f159bc6e"} -->

Mark-Simulacrum commented:
dev-static kicked off.

Mark-Simulacrum commented:
@rust-timer build 16d56136c42dabd7da968fcd4a16adc8f159bc6e (Just as another spot check)

@rustbot author

r=me with https://github.com/rust-lang/rust/pull/102565#discussion_r1008743286 fixed (if you want), and some trivial testing (e.g., rustup-toolchain-install-master; rustup update with dev-static) done.

rust-timer commented:
Queued 16d56136c42dabd7da968fcd4a16adc8f159bc6e with parent 68c836a904e5a421712db311421c5266f9ce71c0, future [comparison URL](https://perf.rust-lang.org/compare.html?start=68c836a904e5a421712db311421c5266f9ce71c0&end=16d56136c42dabd7da968fcd4a16adc8f159bc6e).

rust-timer commented:
Finished benchmarking commit (16d56136c42dabd7da968fcd4a16adc8f159bc6e): [comparison URL](https://perf.rust-lang.org/compare.html?start=68c836a904e5a421712db311421c5266f9ce71c0&end=16d56136c42dabd7da968fcd4a16adc8f159bc6e&stat=instructions:u).

### Overall result: ✅ improvements - no action needed


Benchmarking this pull request likely means that it is perf-sensitive, so we're automatically marking it as not fit for rolling up. While you can manually mark this PR as fit for rollup, we strongly recommend not doing so since this PR may lead to changes in compiler perf.

@bors rollup=never
@rustbot label: +S-waiting-on-review -S-waiting-on-perf -perf-regression


### [Instruction count](https://perf.rust-lang.org/compare.html?start=68c836a904e5a421712db311421c5266f9ce71c0&end=16d56136c42dabd7da968fcd4a16adc8f159bc6e&stat=instructions:u)
This is a highly reliable metric that was used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -     | -              | 0     |
| Improvements ✅ <br /> (secondary) | -1.1% | [-1.1%, -1.1%] | 1     |
| All ❌✅ (primary)                 | -     | -              | 0     |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=68c836a904e5a421712db311421c5266f9ce71c0&end=16d56136c42dabd7da968fcd4a16adc8f159bc6e&stat=max-rss)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | 3.7%  | [3.3%, 4.3%]   | 3     |
| Improvements ✅ <br /> (primary)   | -1.0% | [-1.0%, -1.0%] | 1     |
| Improvements ✅ <br /> (secondary) | -     | -              | 0     |
| All ❌✅ (primary)                 | -1.0% | [-1.0%, -1.0%] | 1     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=68c836a904e5a421712db311421c5266f9ce71c0&end=16d56136c42dabd7da968fcd4a16adc8f159bc6e&stat=cycles:u)
This benchmark run did not return any relevant results for this metric.

jyn514 committed:
Remove unnecessary argument to `package`

Mark-Simulacrum commented:
Per https://github.com/rust-lang/rust/pull/102565#issuecomment-1296307847 I think this is waiting on some additional testing which should hopefully be possible on your end? Let me know if not.

@rustbot author

jyn514 commented:
Looks pretty good to me :) 
```
; RUSTUP_DIST_SERVER=https://dev-static.rust-lang.org rustup toolchain install nightly-2022-10-29
info: syncing channel updates for 'nightly-2022-10-29-x86_64-unknown-linux-gnu'
info: latest update on 2022-10-29, rust version 1.66.0-nightly (16d56136c 2022-10-29)
info: downloading component 'cargo'
info: downloading component 'clippy'
info: downloading component 'rust-docs'
 19.0 MiB /  19.0 MiB (100 %)   9.2 MiB/s in  2s ETA:  0s
info: downloading component 'rust-std'
 29.6 MiB /  29.6 MiB (100 %)  11.2 MiB/s in  3s ETA:  0s
info: downloading component 'rustc'
 67.9 MiB /  67.9 MiB (100 %)  11.2 MiB/s in  7s ETA:  0s
info: downloading component 'rustfmt'
info: installing component 'cargo'
  3 IO-ops /   3 IO-ops (100 %)   2 IOPS in  1s ETA:  0s    
info: installing component 'clippy'
info: installing component 'rust-docs'
 19.0 MiB /  19.0 MiB (100 %) 875.2 KiB/s in  1m  7s ETA:  0s
info: installing component 'rust-std'
 29.6 MiB /  29.6 MiB (100 %)   6.4 MiB/s in  5s ETA:  0s
 13 IO-ops /  13 IO-ops (100 %)   0 IOPS in  6s ETA: Unknown
info: installing component 'rustc'
 67.9 MiB /  67.9 MiB (100 %)  15.6 MiB/s in  4s ETA:  0s
  6 IO-ops /   6 IO-ops (100 %)   0 IOPS in 26s ETA: Unknown
info: installing component 'rustfmt'
  1 IO-ops /   1 IO-ops (100 %)   0 IOPS in  1s ETA: Unknown

  nightly-2022-10-29-x86_64-unknown-linux-gnu installed - rustc 1.66.0-nightly (16d56136c 2022-10-29)

info: checking for self-updates
; cargo +nightly-2022-10-29-x86_64-unknown-linux-gnu --version
cargo 1.66.0-nightly (7e484fc1a 2022-10-27)
; rustc +nightly-2022-10-29-x86_64-unknown-linux-gnu --version
rustc 1.66.0-nightly (16d56136c 2022-10-29)
```

jyn514 commented:
@bors r=Mark-Simulacrum

bors commented:
:pushpin: Commit 8810174a16a72d6fa4bf1322adb54acda0f40b27 has been approved by `Mark-Simulacrum`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=Mark-Simulacrum 8810174a16a72d6fa4bf1322adb54acda0f40b27 -->
<!-- homu: {"type":"Approved","sha":"8810174a16a72d6fa4bf1322adb54acda0f40b27","approver":"Mark-Simulacrum","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:hourglass: Testing commit 8810174a16a72d6fa4bf1322adb54acda0f40b27 with merge 0aaad9e757207657064d70bf9e6c1e6eb327bf15...
<!-- homu: {"type":"BuildStarted","head_sha":"8810174a16a72d6fa4bf1322adb54acda0f40b27","merge_sha":"0aaad9e757207657064d70bf9e6c1e6eb327bf15"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/3427313823/jobs/5712878204)
Approved by: Mark-Simulacrum
Pushing 0aaad9e757207657064d70bf9e6c1e6eb327bf15 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"Mark-Simulacrum","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/3427313823/jobs/5712878204"},"merge_sha":"0aaad9e757207657064d70bf9e6c1e6eb327bf15"} -->

rust-timer commented:
Finished benchmarking commit (0aaad9e757207657064d70bf9e6c1e6eb327bf15): [comparison URL](https://perf.rust-lang.org/compare.html?start=d22c43389da86abba64fbed8b4a7e2ba8a1a7c10&end=0aaad9e757207657064d70bf9e6c1e6eb327bf15&stat=instructions:u).

### Overall result: no relevant changes - no action needed


@rustbot label: -perf-regression



### [Instruction count](https://perf.rust-lang.org/compare.html?start=d22c43389da86abba64fbed8b4a7e2ba8a1a7c10&end=0aaad9e757207657064d70bf9e6c1e6eb327bf15&stat=instructions:u)
This benchmark run did not return any relevant results for this metric.

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=d22c43389da86abba64fbed8b4a7e2ba8a1a7c10&end=0aaad9e757207657064d70bf9e6c1e6eb327bf15&stat=max-rss)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | 2.7%  | [2.3%, 3.1%]   | 4     |
| Improvements ✅ <br /> (primary)   | -     | -              | 0     |
| Improvements ✅ <br /> (secondary) | -3.9% | [-3.9%, -3.9%] | 1     |
| All ❌✅ (primary)                 | -     | -              | 0     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=d22c43389da86abba64fbed8b4a7e2ba8a1a7c10&end=0aaad9e757207657064d70bf9e6c1e6eb327bf15&stat=cycles:u)
This benchmark run did not return any relevant results for this metric.
